### ISSUE TITLE
Unexpected behaviour with stdin and .lines() on Windows

### ISSUE THREAD

**Issue by DaniJVaz** [rust-lang/rust#26710]
Using .lines() for input on Windows always reads an '\r' at the end of the string.

For example:

	fn main() {
		use std::io::prelude::*;
		use std::io;

		let stdin = io::stdin();
		for line in stdin.lock().lines() {
			let line = line.unwrap();
			println!("You entered (String): {:?}", line);

			match line.parse::<i32>() {
				Ok(num) => println!("You entered (i32): {}", num),
				Err(_) => println!("That was not a number..."),
			}
		}
	}

prints:

	2
	You entered (String): "2\r"
	That was not a number...

This complicates even simple tasks, like reading a number in the example, which now needs an extra .trim() to make sure that the '\r' is removed. 

I understand that in some cases having the '\r' may be useful, but when using .lines() the input should be split into lines according to the line separator, which is "\r\n" on Windows.

**Comment by RalfJung** [comment]
I should add that the code above works fine on Linux (with Rust 1.2 Beta). So this is a cross-platform compatibility hazard.

**Comment by arielb1** [comment]
This is just "binary mode".

**Comment by steveklabnik** [comment]
Yes, this is expected behavior. You get a `\n` on the end in Linux as well.

**Comment by DaniJVaz** [comment]
The documentation explicitly says that this shouldn't happen

> Each string returned will *not* have a newline byte (the 0xA byte) at the end.
https://doc.rust-lang.org/std/io/trait.BufRead.html#method.lines

Additionally, if this was the case, I would think that in Windows it would be '\r\n' instead of just '\r'. 

Anyway, if .lines() is supposed to work in "binary mode", then everything is technically fine, though I still think it would be nice to have (maybe optional) "text mode".

**Comment by steveklabnik** [comment]
If this is true, it's been a bug for a very long time, as the Guessing Game example has used `trim()` for this purpose for.. a year?

Let's ask @rust-lang/libs if this behavior or the documentation is wrong.

**Comment by RalfJung** [comment]
@steveklabnik The guessing game does not use the `lines()` iterator, it uses `read_line()`. That function is documented to include the delimiter, while `lines()` is documented to *not* include it. And reality matches the documentation on Linux: the `\n` is stripped. On Windows, the `\n` is stripped as well, but the `\r`is not. That's very surprising.

After all, there's a reason many languages have a "text mode".

**Comment by steveklabnik** [comment]
Ugh, you are correct. I should finish coffee before starting my morning triage :(

This is pretty certainly a bug, then. Sorry :/

**Comment by arielb1** [comment]
That's not a Windows/Linux issue - you can have files with DOS-style line-endings on Linux and files with Unix-style line-endings on Windows.

**Comment by alexcrichton** [comment]
This is currently working as intended and it's also why `lines_any` exists on strings, and we can perhaps investigate adding `lines_any` to the `Read` trait.

**Comment by RalfJung** [comment]
Wow, that's certainly surprising. I thought Rust was supposed to abstract away from the ugly platform differences. Why are character sets more important to abstract away from than line endings? At the very least, I'd argue this warrants a *big warning* in the documentation.

`lines_any` on a `String` is obviously not very helpful in that situation. Should I write a feature request for `line_any` on `Read` (or rather `BufRead`, I suppose?). I kind of feel like giving that function a try myself ;-)

**Comment by arielb1** [comment]
@RalfJung 

DOS/Unix line-endings aren't a platform difference – DOS line-endings are more common on Windows and Unix line-endings are more common on Linux, but you can get both kinds of line-endings on both platforms (e.g. files can be transmitted between machines).

**Comment by RalfJung** [comment]
In general, I agree. For the specific case of "reading from interactive stdin" though, this reduces to a platform difference.

**Comment by bluss** [comment]
Shouldn't `.lines()` just strip `\n` or `\r\n` if they are present? The current `.lines()` seems to a simple "dwim" method. It's surprising to me that it doesn't do this since we chose to have a trimming lines iterator. It seems unfair to burden the user.

**Comment by retep998** [comment]
I really think that anywhere we strip `\n` we should also strip `\r\n`.

**Comment by RalfJung** [comment]
I submitted a PR for implementing `lines_any` on `BufRead`: <https://github.com/rust-lang/rust/pull/26743>. However, I agree with @retep998 and @bluss in that the expectation of a method called `lines()` is to support both kinds of line endings. I don't even see any reason *not* to do so. A programmer that explicitly wants to split only at `\n` should use a byte-based splitting function.

However, I don't have any idea whether changing the behavior of `BufRead::lines` and `str::lines` accordingly is compatible with Rust's stability guarantees.

**Comment by RalfJung** [comment]
The pull request <https://github.com/rust-lang/rust/pull/26743> is now in final comment period, so that's the right place to comment about "adding lines_any to BufRead" vs. "changing lines to cover Windows line endings".

**Comment by alexcrichton** [comment]
I'm going to close this in favor of the associated RFC: https://github.com/rust-lang/rfcs/pull/1212
